packages/flutter_test/lib/src/finders.dart

```diff
@@ -460,57 +469,544 @@ class CommonFinders {
+/// for searching Flutter's element and semantics trees respectively. If the
+/// search can be represented as a predicate, then use [MatchFinderMixin] along
+/// with the appropriate base class.
+abstract class FinderBase<TCandidate> {
+  /// The results of the latest [find] or [tryFind] call.
+  ///
+  /// Unlike [find] and [tryFind], [found] will not re-execute the search for
+  /// this finder. Either [find] or [tryFind] must be called before accessing
+  /// [found].
+  FinderResult<TCandidate> get found {
+    assert(
+      _found != null,
+      'No results have been found yet. '
+      'Either `find` or `tryFind` must be called before accessing `found`',
+    );
+    return _found!;
+  }
+  FinderResult<TCandidate>? _found;
+
+  /// Whether or not this finder has any results in [found].
+  bool get hasFound => _found != null;
+
+  /// Describes what the finder is looking for.
+  ///
+  /// {@template flutter_test.finders.FinderBase.requirements}
+  /// This string should be a brief English phrase describing the finder's
+  /// requirements for a single candidate. It should make sense as the end of
+  /// the sentence "Found an item <requirements>".
+  ///
+  /// As an example, a `FinderBase<Object>` that searches for a candidate of a
+  /// specific type would provide 'of type MyObject', so that the example
+  /// sentence comes out to: "Found an item of type MyObject."
+  /// {@endtemplate}
+  ///
+  /// See also:
+  ///
+  ///   * [FinderBase.toString] where this is used to fully describe the finder
+  ///   * [FinderResult.toString] where this is used to provide context to the
+  ///     results of a search
+  String get requirements;
+
+  /// Returns all of the items that will be considered by this finder.
+  @protected
+  Iterable<TCandidate> get allCandidates;
+
+  /// Returns a variant of this finder that only matches the first item
+  /// found by this finder.
+  FinderBase<TCandidate> get first => _FirstFinder<TCandidate>(this);
+
+  /// Returns a variant of this finder that only matches the last item
+  /// found by this finder.
+  FinderBase<TCandidate> get last => _LastFinder<TCandidate>(this);
+
+  /// Returns a variant of this finder that only matches the item at the
+  /// given index found by this finder.
+  FinderBase<TCandidate> at(int index) => _IndexFinder<TCandidate>(this, index);
+
+  /// Returns all the items in the given list that match this
+  /// finder's requirements.
+  ///
+  /// When implementing your own finders that inherit directly from
+  /// [FinderBase], this is the main method to override. If your finder
   /// can efficiently be described just in terms of a predicate
-  /// function, consider extending [MatchFinder] instead.
-  Iterable<Element> apply(Iterable<Element> candidates);
+  /// function, consider mixing in [MatchFinderMixin] instead.
+  @protected
+  Iterable<TCandidate> findInCandidates(Iterable<TCandidate> candidates);
```

Comment:
Just for my own understanding: why is this method given `candidates` instead of just making it operate on `allCandidates`?

Reply:
The `allCandidates` property has side effects (evaluating the element/semantics trees), so we want the `find` method to closely control the number of calls to it. This method is the one that's intended to be overridden instead. That idea was actually taken from the original Finder API.